### PR TITLE
add i2c trait

### LINKED ISSUE #51: I2C traits
I'm extensively using I2C in my project together with `embassy` and it would be nice to have an I2C interface trait to implement for a particular chip (stm32wb55 in my case).

### PR DESCRIPTION
**Pull request by xoviat**
closes #51.

### CODE REVIEW

**Comment by xoviat** [comment]
mostly copied from embedded-hal; transactional mod can be sorted later

File: embassy-traits/src/i2c.rs [hunk 1]
```diff
@@ -0,0 +1,409 @@
+//! Blocking I2C API
```
Comment (eupn) [inline embassy-traits/src/i2c.rs:1]: Copy-paste here.

File: embassy-traits/src/i2c.rs [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+//!
+//! Additionally, the I2C 10-bit address mode has been developed to be fully
+//! backwards compatible with the 7-bit address mode. This allows for a
+//! software-emulated 10-bit addressing implementation if the address mode
+//! is not supported by the hardware.
+//!
+//! Since 7-bit addressing is the mode of the majority of I2C devices,
+//! `SevenBitAddress` has been set as default mode and thus can be omitted if desired.
+//!
+//! ## Examples
```
Comment (eupn) [inline embassy-traits/src/i2c.rs:19]: Code examples should be either updated or removed.

File: embassy-traits/src/i2c.rs [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+    /// Where
+    ///
+    /// - `ST` = start condition
+    /// - `SAD+R` = slave address followed by bit 1 to indicate reading
+    /// - `SAK` = slave acknowledge
+    /// - `Bi` = ith byte of data
+    /// - `MAK` = master acknowledge
+    /// - `NMAK` = master no acknowledge
+    /// - `SP` = stop condition
+    fn read<'a>(&mut self, address: A, buffer: &mut [u8]) -> Self::ReadFuture<'a>;
```
Comment (eupn) [inline embassy-traits/src/i2c.rs:147]: As per @Dirbaio's suggestion, we should add a `Pin` restriction to `self`: `self: Pin<&mut self>` for these traits. This should allow implementations to handle interrupts from I2C/DMA peripherals.

**Comment by Dirbaio** [comment]
I'm not sure about the design. There are *many* traits: Read, Write, WriteIter, WriteRead, WriteIterRead, Transactional.

- What's the advantage in separating Read, Write, WriteRead? I don't think there's hardware that supports only reading but not writing or viceversa.

- WriteIter, WriteIterRead don't seem very useful to me if we're going to implement this with DMA. It forces impls to write byte-by-byte.

The Transactional one is nice to have :D Unfortunately it's [not possible](https://devzone.nordicsemi.com/f/nordic-q-a/66615/nrf52840-twim-how-to-write-multiple-buffers-without-repeated-start-condition) to implement it on nrf52, but I imagine it can be useful in other contexts.

**Comment by eupn** [comment]
@Dirbaio as far as `Read` trait concerned, there are chips (BNO08x for example) that use I2C only as a transport (like uart, etc.) for bytes and doesn’t rely on first writing a register address before reading data.

**Comment by Dirbaio** [comment]
Yeah, having `read`, `write`, `write_read` methods make sense. I was wondering why is it necessary to have 3 traits, instead of 1 trait with 3 methods.

Also, `read` and `write` could really be just `write_read` with zero-size slices, so having just a `write_read` method could work too

**Comment by xoviat** [comment]
This is copied from embedded-hal. Why are there different traits there?

**Comment by Dirbaio** [comment]
I did some archeology on embedded-hal, only thing I could find is [this](https://github.com/rust-embedded/embedded-hal/issues/2#issuecomment-307512572). The [commit](https://github.com/rust-embedded/embedded-hal/commit/5415ee164bb3f4e178910420da42d1287b74e7df) that adds them already has them this way.

Also from Matrix:

```
Dirbaio: Does anyone know why embedded-hal traits are always 1 method per trait?
Dirbaio: For example: for i2c theres Read, Write, WriteRead traits with one method each. 
jamesmunns: Generally more flexibility
Dirbaio: what's the advantage vs having 1 trait with the 3 methods? 
Dirbaio: I kinda see it with uart read/write, you may want to "split" it
Dirbaio: but that doesn't make sense at all with i2c
jamesmunns: If you have hardware that can only do 2 of those three things, it can work with drivers that don't need that thing?
Dirbaio: is there hardware that can only i2c-write and can't i2c read? 
Dirbaio: or viceversa?
jamesmunns: I mean
jamesmunns: Not i2c, but spi yeah
jamesmunns: Like a tx only spi port for ws2812b spi
Dirbaio: hm, right, for uart/spi it does make sense
jamesmunns: Then probably just consistency
```

It's a good point for UART and SPI, but i'm not very convinced it's good to split I2C (neither for the errors nor for consistency)

**Comment by Dirbaio** [comment]
Don't delete write_read. It's not the same as write + read: It does a Repeated Start, instead of Stop+Start.

**Comment by Dirbaio** [comment]
:+1:

**Pull request merged**